### PR TITLE
ec2-agent/vsphere-vm-agent: allow custom userdata

### LINKED ISSUE #627: ec2-agent: Allow custom userdata when launching instances
Allow the user to provide custom user-data TOML that will be appended/merged into the user-data settings for test nodes created by the ec2 resource agent.

### PR DESCRIPTION
**Pull request by mjsterckx**
**Issue number:**

Closes #627 

**Description of changes:**

The `Ec2Config` and `VSphereVmConfig` are modified to allow an optional `CustomUserData` field. This field contains either a `Merge` or a `Replace` value with encoded userdata. The userdata either replaces the default userdata (if the `Replace` mode is used) or is parsed into `toml::Value` and merged with the default userdata (if the `Merge` mode is used).

**Testing done:**

- Unit tests for `merge_values`
- Ran an ECS test with this TOML file as custom userdata (in merge mode):
```
[settings.ecs]
loglevel = "warn"

[settings]
motd = "Custom userdata"
```

Test passed and instance userdata looks like this:
```
[settings]
motd = "Custom userdata"

[settings.ecs]
cluster = "external-cluster"
loglevel = "warn"
```
- Ran an EKS test with this TOML file as custom userdata (in merge mode):
```
[settings.kubernetes]
log-level = 3

[settings]
motd = "Custom userdata"
```

Test passed and instance userdata looks like this:
```
[settings]
motd = "Custom userdata"

[settings.kubernetes]
api-server = "https://ECEE40E694F1EA9026E6415225549739.gr7.us-east-1.eks.amazonaws.com"
cluster-certificate = "****************"
cluster-dns-ip = "10.100.0.10"
cluster-name = "eks-test-cluster"
log-level = 3

[settings.updates]
ignore-waves = true
```
- Ran a VMware test with this TOML file as custom userdata (in merge mode):
```
[settings.updates]
ignore-waves = true

[settings]
motd = "Custom userdata"
```

Test passed and userdata looks like this:
```
[settings]
motd = "Custom userdata"
[settings.host-containers.control]
enabled = true
user-data = "eyJzc20iOnsiYWN0aXZhdGlvbi1pZCI6IjY5YmY5YzgzLWEwYTktNDA5My04M2M2LTQxM2FhZTRiNDI1ZiIsImFjdGl2YXRpb24tY29kZSI6IlpyOGJ4ZVA5Q1h6SzJSZ0ZsZXlsIiwicmVnaW9uIjoidXMtd2VzdC0yIn19"

[settings.kubernetes]
api-server = "https://***.**.**.***:****/"
bootstrap-token = "****************"
cluster-certificate = "****************"
cluster-dns-ip = "**.**.*.**"

[settings.updates]
ignore-waves = true
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs
Comment (webern) [inline bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs]: Can you move these string formats out into two functions so that the branching logic here is easier to read?

```rust
fn default_eks_userdata(x, y, z) -> String { todo!() }
fn default_ecs_userdata(x) -> String { todo!() }
```

File: bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs
Comment (webern) [inline bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs]: I would simplify this match case by early returning if there is no custom userdata. Something like 

```rust
let default_userdata = match cluster_type {
    ClusterType::Eks => default_eks_userdata(x, y, z),
    ClusterType::Ecs => default_ecs_userdata(x),
}

let custom_userdata = if let Some(value) = customer_user_data {
    value
} else {
    return Ok(default_userdata);
};

match custom_user_data {
    /* ... */
}
```

Note: in the case of `Replace` the `default_userdata` will have been constructed unnecessarily, but the easy-to-read program flow outweighs the construction of an unused string here.

File: bottlerocket/agents/src/bin/vsphere-vm-resource-agent/vsphere_vm_provider.rs
Comment (webern) [inline bottlerocket/agents/src/bin/vsphere-vm-resource-agent/vsphere_vm_provider.rs]: We could use the same program flow and refactor here that I described for EKS.

File: bottlerocket/agents/src/userdata.rs
Comment (webern) [inline bottlerocket/agents/src/userdata.rs]: I prefer left-to-right, `from, to` ordering (I always mess up when creating symlinks `ln -s to from`)

**Review by webern — COMMENTED** [review]
I see you are pushing changes so I will save what I've commented on so far. Nice work.

File: bottlerocket/agents/src/userdata.rs
Comment (webern) [inline bottlerocket/agents/src/userdata.rs]: it's  safest to check again in case somebody deletes the code above. It's easiest to just say "the code must never panic" than to try to reason about `unwrap` statements. (then again, this isn't avionics, so :man_shrugging: )

```suggestion
            let t1 = merge_into.as_table_mut().ok_or_else(|| IntoProviderError::context(
            None,
            Resources::Clear,
            "Cannot merge mismatched data types in given TOML",
        ))?;
```

Comment (webern) [inline bottlerocket/agents/src/userdata.rs]: Easier to read if we know which one is from/to. 

```suggestion
        Value::Table(from) => {
            // We know the other side is a table because of the `ensure` above.
            let to = merge_into.as_table_mut().unwrap();
```

File: bottlerocket/agents/src/userdata.rs
Comment (webern) [inline bottlerocket/agents/src/userdata.rs]: Woah is this my fault? Should this be named IntoProviderResult instead? (don't change it in this PR, I am just surprised by this function body)

File: bottlerocket/agents/src/userdata.rs
Comment (webern) [inline bottlerocket/agents/src/userdata.rs:76]: Thank you for writing a test. You could probably add a couple more to cover cases such as
- left has the key, right does not
- right has the key, left does not
- error cases

File: bottlerocket/testsys/src/run_aws_ecs.rs
Comment (webern) [inline bottlerocket/testsys/src/run_aws_ecs.rs]: We shouldn't add two optional fields to a struct that must both be `Some` in order for the structure to be in a consistent state. Instead we should have

```rust
struct CustomUserDataConfig {
    data: String,
    mode: CustomUserDataMode,
}


```

Then in the main struct:

```
    /// The path to a TOML file containing custom userdata.
    #[structopt(long, requires("custom_user_data_mode"))]
    custom_user_data: Option<CustomUserData>,
```

I know @ecpullen doesn't like this "tagging"-patterned use of an enum, but in this case since it has to work with yaml it might be the only way to make it self-consistent and self-documenting that both fields are required.

File: bottlerocket/testsys/src/run_aws_ecs.rs
Comment (webern) [inline bottlerocket/testsys/src/run_aws_ecs.rs]: This is the error we should avoid. Especially since the "invalid custom userdata mode" might be `Option::None`.

If you add a "tagging" enum like I requested above, which I called `CustomerUserdataMode`, you could:

```rust
impl CustomUserData {
    fn new(x: CustomUserDataConfig) -> Self {
       match x.mode {
           /* ... */
       }
   }
}
```

File: bottlerocket/testsys/src/run_aws_k8s.rs
Comment (webern) [inline bottlerocket/testsys/src/run_aws_k8s.rs]: same thing, avoid designing structures that are not self-validating. i.e. struct.field_a should not rely on struct.field_b being in a compatible state.

File: bottlerocket/testsys/src/run_aws_k8s.rs
Comment (webern) [inline bottlerocket/testsys/src/run_aws_k8s.rs]: repeated code. maybe my type suggestions above would eliminate it.

File: bottlerocket/testsys/src/run_vmware.rs
Comment (webern) [inline bottlerocket/testsys/src/run_vmware.rs]: same comment about struct fields not relying on the state of other struct fields being consistent

File: bottlerocket/testsys/src/run_vmware.rs
Comment (webern) [inline bottlerocket/testsys/src/run_vmware.rs]: looks like repeated code,

**Comment by mjsterckx** [comment]
^ Addressed some of webern's comments.

File: bottlerocket/testsys/src/error.rs
Comment (webern) [inline bottlerocket/testsys/src/error.rs]: Throughout our projects we format errors like this:
- When adding some input information we surround it with single quotes
- When including an underlying error we put it at the end after a colon, without single quotes.

So your error should be

```suggestion
    #[snafu(display("Error parsing TOML file '{}'", path))]
```
Ideally you would have an underlying error for context and it would look like this:

```
    #[snafu(display("Error parsing TOML file '{}': {}", path, source))]
```

**Review by webern — COMMENTED** [review]
I see you are pushing changes so I will save what I've commented on so far. Nice work.

File: bottlerocket/testsys/src/run_aws_ecs.rs
Comment (mjsterckx) [reply bottlerocket/testsys/src/run_aws_ecs.rs]: Doesn't the `requires` `structopt` ensure that either none of them are present, or both of them are?

That's how it's used in `cli` anyway.

Comment (mjsterckx) [reply bottlerocket/testsys/src/run_aws_ecs.rs:164]: That's how it's used in `cli` anyway.

**Comment by mjsterckx** [comment]
^ Added `source` field to `ParseToml` error type

File: bottlerocket/testsys/src/run_aws_ecs.rs
Comment (webern) [reply bottlerocket/testsys/src/run_aws_ecs.rs]: > Doesn't the requires structopt 

It might, but that doesn't prevent you from writing gross code when you go to use the fields and have to handle the inconsistent-state cases anyway (because they are Options and this is Rust)...

File: bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs
Comment (ecpullen) [inline bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs:360]: Does the user data need to be decoded?

File: bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs
Comment (ecpullen) [inline bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs:388]: Is there a reason not to keep this newline?

File: bottlerocket/agents/src/userdata.rs
Comment (ecpullen) [inline bottlerocket/agents/src/userdata.rs]: I would prefer to use the existing lib in Bottlerocket instead of duplicating all of the code.

File: bottlerocket/testsys/src/run_aws_ecs.rs
Comment (ecpullen) [reply bottlerocket/testsys/src/run_aws_ecs.rs]: `testsys` is going away, so I ok with any solution (even leaving it out).

Comment (webern) [reply bottlerocket/testsys/src/run_aws_ecs.rs]: Sorry it took me a minute to realize this was an Args struct. You can ignore this comment thread.

File: bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs
Comment (mjsterckx) [reply bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs:388]: The newline is still there, I just removed a tab that must've been accidentally added at some point.

File: bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs
Comment (mjsterckx) [reply bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs:360]: The TOML file is read as string and contained in `encoded_userdata` so I don't think so? But I suppose I should change the name of that variable.

Comment (ecpullen) [reply bottlerocket/agents/src/bin/ec2-resource-agent/ec2_provider.rs:360]: The userdata should be base64 encoded.

**Comment by mjsterckx** [comment]
^ Addressed several comments: switched `from`, `to` order in `merge_values`; replaced `unwrap` in `merge_values` with a second check to see if the types match; renamed `t2` and `t1` to `from` and `to` respectively; added code to the CLI to encode userdata as base64 and added a method to `userdata.rs` to decode back to a String.

File: bottlerocket/agents/src/userdata.rs
Comment (mjsterckx) [reply bottlerocket/agents/src/userdata.rs]: `IntoProviderResult` does not exist (`IntoProviderError` does, but that's another matter). This function was removed, however, so I am marking this conversation as resolved.

**Comment by mjsterckx** [comment]
^ Added a couple more unit tests (key present on left side but not on right; mismatched type error)

**Comment by mjsterckx** [comment]
^ Small fixes related to encoding/decoding

File: bottlerocket/agents/src/userdata.rs
Comment (etungsten) [inline bottlerocket/agents/src/userdata.rs]:
```suggestion
// This is only used in the build script for now, but it's in a separate crate so we can have unit
// tests.
```
I don't think this comment applies here.

After looking through this, I think we should instead switch to using the `merge-toml` library from the bottlerocket repo directly https://github.com/bottlerocket-os/bottlerocket/blob/67e5abfb0e1575fb6bcb7a710cdb3d422ef12283/sources/api/storewolf/merge-toml/src/lib.rs like @ecpullen mentioned before. We can still use the custom error types and keep `decode_to_string` somewhere else/

**Review by etungsten — CHANGES_REQUESTED** [review]

File: bottlerocket/agents/src/userdata.rs
Comment (mjsterckx) [reply bottlerocket/agents/src/userdata.rs]: I can remove the comment, but copying the code instead of using the library is something that was discussed and @webern especially felt like we should not rely on a library from a different project that was never meant to be included in other projects.

Comment (webern) [reply bottlerocket/agents/src/userdata.rs]: This is true. I am hesitant to create the first ever external dependency on library code that resides in Bottlerocket. I'd rather we take an action item to create a merge_toml crate external to both projects and put it on crates.io than to have dependencies going in both directions (Bottlerocket depends on testsys which depends on Bottlerocket).

Comment (etungsten) [reply bottlerocket/agents/src/userdata.rs]: Can you add a comment to add some attribution for this code? Just to keep track that it's copied from bottlerocket. Adding the permalink and saying this is directly borrow from there is probably enough. https://github.com/bottlerocket-os/bottlerocket/blob/develop/sources/api/storewolf/merge-toml/src/lib.rs

See https://github.com/bottlerocket-os/bottlerocket/blob/f915de1a074d88a16348b444d31c49aaff51f8e9/sources/host-ctr/cmd/host-ctr/registry.go#L48-L52 for an example.

Comment (ecpullen) [reply bottlerocket/agents/src/userdata.rs]: Can you also create an issue in both repos so we make sure this gets fixed.

File: bottlerocket/testsys/src/run_aws_k8s.rs
Comment (ecpullen) [inline bottlerocket/testsys/src/run_aws_k8s.rs]: Let's make this an enum.

File: bottlerocket/testsys/src/run_aws_k8s.rs
Comment (ecpullen) [inline bottlerocket/testsys/src/run_aws_k8s.rs]:
```suggestion
let user_data = &self.custom_user_data
        .map(|path| read_to_string(path.clone()))
        .transpose()
        .context(error::ReadSnafu{})?
        .map(base64::encode);
        let user_data = match (self.custom_user_data_mode, user_data){
                  (Some(_),None)|(None,Some(_)) => return Err(error::Error::InvalidArguments { why: "Either both or neither of custom-user-data-mode and custom-user-data must be provided.".to__string() }),
                 (Some(UserdataType::Merge),Some(userdata)) => Some(CustomUserData::Merge{encoded_userdata: userdata}),
                 (Some(UserdataType::Replace),Some(userdata)) => Some(CustomUserData::Replace{encoded_userdata: userdata}),
                 (None,None) => None
        };
```

**Review by webern — APPROVED** [review]
Looks good when the erikson's and ethan's suggestions are implemented. Nice work.

**Comment by mjsterckx** [comment]
^ Addressed ecpullen's suggestion and added attribution to indicate that `userdata.rs` is mostly copied from the Bottlerocket repo.

File: bottlerocket/agents/src/userdata.rs
Comment (etungsten) [inline bottlerocket/agents/src/userdata.rs]: quick nit: sorry, this should be a more permanent link that's tagged:
```suggestion
// See https://github.com/bottlerocket-os/bottlerocket/blob/v1.11.1/sources/api/storewolf/merge-toml/src/lib.rs
```

**Review by etungsten — APPROVED** [review]

**Review by etungsten — APPROVED** [review]

File: bottlerocket/agents/src/userdata.rs
Comment (ecpullen) [inline bottlerocket/agents/src/userdata.rs:73]: I'm not sure this belongs here, but I won't block on it.

**Review by ecpullen — APPROVED** [review]

File: bottlerocket/agents/src/userdata.rs
Comment (mjsterckx) [reply bottlerocket/agents/src/userdata.rs:73]: For now, I think it's fine keeping it in a file called `userdata` since the function is only used in context of the userdata. Once the `merge-toml` code gets imported as library instead of being copied, this function can be moved to a `utils` location.

**Pull request merged**
